src/Workspaces/Core/Portable/Rename/IRenameRewriterLanguageService.cs

```diff
@@ -149,5 +157,143 @@ internal abstract class AbstractRenameRewriterLanguageService : IRenameRewriterL
                 }
             }
         }
+
+        protected static Dictionary<SymbolKey, RenameSymbolContext> GroupRenameContextBySymbolKey(
+            ImmutableArray<RenameSymbolContext> symbolContexts, IEqualityComparer<SymbolKey> comparer)
+        {
+            var renameContexts = new Dictionary<SymbolKey, RenameSymbolContext>(comparer);
+            foreach (var context in symbolContexts)
+            {
+                renameContexts[context.RenamedSymbol.GetSymbolKey()] = context;
+            }
+
+            return renameContexts;
+        }
+
+        protected static Dictionary<TextSpan, LocationRenameContext> GroupTextRenameContextsByTextSpan(
+            ImmutableArray<LocationRenameContext> textSpanRenameContexts)
+        {
+            var textSpanToRenameContext = new Dictionary<TextSpan, LocationRenameContext>();
+            foreach (var context in textSpanRenameContexts)
+            {
+                var textSpan = context.RenameLocation.Location.SourceSpan;
+                if (!textSpanToRenameContext.TryGetValue(textSpan, out var existingRenameContext))
+                {
+                    textSpanToRenameContext[textSpan] = context;
+                }
+                else if (!existingRenameContext.Equals(context))
+                {
+                    // A textSpan is being renamed with different rename location info.
+                    throw new ArgumentException(
+                        $"{textSpan} of {context.RenameLocation.DocumentId} is being renamed to two differnt locations info: {existingRenameContext}, {context}. ");
+                }
+            }
+
+            return textSpanToRenameContext;
+        }
+
+        protected static Dictionary<TextSpan, HashSet<LocationRenameContext>> GroupStringAndCommentsTextSpanRenameContexts(
```

Comment:
Do these need to return `ImmutableDictionary` instead?

Reply:
ImmutableDictionary might be a littble bit slower. But I am fine with starting with the Immuable one